File: apps/desktop/src/bridge/bridgeHostStatus.ts
import { invoke } from "@tauri-apps/api/core";

export type BridgeHostStatus = "running" | "port_conflict" | "failed" | "unknown";

export async function getBridgeHostStatus(): Promise<BridgeHostStatus> {
  try {
    const status = await invoke<string>("get_bridge_host_status");

    if (
      status === "running" ||
      status === "port_conflict" ||
      status === "failed"
    ) {
      return status;
    }
  } catch {
    return "unknown";
  }

  return "unknown";
}

export function formatBridgeHostStatus(status: BridgeHostStatus): string {
  if (status === "running") {
    return "running";
  }

  if (status === "port_conflict") {
    return "port conflict on 127.0.0.1:17777";
  }

  if (status === "failed") {
    return "failed";
  }

  return "unknown";
}

File: apps/desktop/src-tauri/src/bridge_host_status.rs
use std::io;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BridgeHostStatus {
    value: &'static str,
}

impl BridgeHostStatus {
    pub fn running() -> Self {
        Self { value: "running" }
    }

    pub fn from_start_error(error: &io::Error) -> Self {
        if error.kind() == io::ErrorKind::AddrInUse {
            return Self {
                value: "port_conflict",
            };
        }

        Self { value: "failed" }
    }

    pub fn as_str(&self) -> &'static str {
        self.value
    }
}

#[tauri::command]
pub fn get_bridge_host_status(status: tauri::State<'_, BridgeHostStatus>) -> String {
    status.as_str().to_string()
}

#[cfg(test)]
mod tests {
    use super::BridgeHostStatus;
    use std::io;

    #[test]
    fn classifies_addr_in_use_as_port_conflict() {
        let error = io::Error::from(io::ErrorKind::AddrInUse);

        assert_eq!(
            BridgeHostStatus::from_start_error(&error).as_str(),
            "port_conflict"
        );
    }

    #[test]
    fn classifies_other_start_errors_as_failed() {
        let error = io::Error::from(io::ErrorKind::PermissionDenied);

        assert_eq!(BridgeHostStatus::from_start_error(&error).as_str(), "failed");
    }
}
